core-kotlin-modules/core-kotlin-advanced-3/src/test/kotlin/com/baeldung/withLockVsSynchronize/WithLockVsSynchronizeUnitTest.kt

```diff
@@ -0,0 +1,93 @@
+package com.baeldung.withLockVsSynchronize
+
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.thread
+import kotlin.concurrent.withLock
+
+
+class WithLockVsSynchronizeUnitTest {
+
+    object ObjectToLockOn
+    val mutexLock = Mutex()
+    @Test
+    fun `test synchronized keyword usage on a class`() {
+        val counter = CounterClass()
+        val threads = List(500) { index ->
+            thread {
+                val value = counter.incrementAndGet()
+                assertEquals(index + 1, value)
+            }
+        }.forEach { it.join() }
+
+        assertEquals(500, counter.getCount())
+    }
+
+    @Test
+    fun `test withLock method usage`() {
+        val counter = LockCounterClass()
+        val threads = List(500) { index ->
+            thread {
+                val value = counter.incrementAndGet()
+                assertEquals(index + 1, value)
+            }
+        }.forEach { it.join() }
+
+        assertEquals(500, counter.getCount())
+    }
+
+    fun toSynchronized() = synchronized(ObjectToLockOn) {
+        println("Synchronized function call")
+    }
+
+    @Synchronized
+    fun synchronizedMethod() {
+        println("Synchronized function call")
+    }
+
+    fun performSynchronizedOperation(){
+        val lockObject = ReentrantLock()
```

Comment:
This lock needs to be declared outside of the function, otherwise it won't actually synchronize anything because we create a new lock with every function call.